crates/recursor/src/tests.rs

```diff
@@ -0,0 +1,201 @@
+            .len();
+        let tld_tcp_conns = calls
+            .iter()
+            .filter(|(ip, proto)| *ip == TLD_IP && *proto == ProtocolConfig::Tcp)
+            .collect::<Vec<_>>()
+            .len();
+        let leaf_tcp_conns = calls
+            .iter()
+            .filter(|(ip, proto)| *ip == LEAF_IP && *proto == ProtocolConfig::Tcp)
+            .collect::<Vec<_>>()
+            .len();
+
+        assert_eq!(root_tcp_conns, 1);
+        assert_eq!(tld_tcp_conns, 1);
+        assert_eq!(leaf_tcp_conns, 1);
+
+        // With the ns_cache_size set to 1, we should see reconnections for the TLD
+        // and leaf queries because the NameServer objects have dropped out of the
+        // connection cache.
+        let response = recursor
+            .resolve(
+                Query::query(dup_query_name, RecordType::A),
+                Instant::now(),
+                false,
+            )
+            .await
+            .unwrap();
+
+        assert_eq!(response.response_code(), ResponseCode::NoError);
+
+        let calls = provider.new_connection_calls();
+        let root_tcp_conns = calls
+            .iter()
+            .filter(|(ip, proto)| *ip == ROOT_IP && *proto == ProtocolConfig::Tcp)
+            .collect::<Vec<_>>()
+            .len();
+        let tld_tcp_conns = calls
+            .iter()
+            .filter(|(ip, proto)| *ip == TLD_IP && *proto == ProtocolConfig::Tcp)
+            .collect::<Vec<_>>()
+            .len();
+        let leaf_tcp_conns = calls
+            .iter()
+            .filter(|(ip, proto)| *ip == LEAF_IP && *proto == ProtocolConfig::Tcp)
+            .collect::<Vec<_>>()
+            .len();
+
+        // Roots aren't subject to cache expiration
+        assert_eq!(root_tcp_conns, 1);
+        assert_eq!(tld_tcp_conns, 2);
+        assert_eq!(leaf_tcp_conns, 2);
+    });
+
+    Ok(())
+}
+
+fn test_fixture() -> Result<(MockProvider, RecursorBuilder<MockProvider>), ProtoError> {
+    let query_name = Name::from_ascii("host.hickory-dns.testing.")?;
+    let dup_query_name = Name::from_ascii("host.hickory-dns-dup.testing.")?;
+
+    let tld_zone = Name::from_ascii("testing.")?;
+    let tld_ns = Name::from_ascii("testing.testing.")?;
+    let leaf_zone = Name::from_ascii("hickory-dns.testing.")?;
+    let leaf_ns = Name::from_ascii("ns.hickory-dns.testing.")?;
+    let dup_leaf_zone = Name::from_ascii("hickory-dns-dup.testing.")?;
+    let dup_leaf_ns = Name::from_ascii("ns.hickory-dns-dup.testing.")?;
+
+    let responses = vec![
+        MockRecord::ns(ROOT_IP, &tld_zone, &tld_ns),
+        MockRecord::a(ROOT_IP, &tld_ns, TLD_IP)
+            .with_query_name(&tld_zone)
+            .with_query_type(RecordType::NS)
+            .with_section(MockResponseSection::Additional),
+        MockRecord::ns(TLD_IP, &leaf_zone, &leaf_ns)
+            .with_query_name(&leaf_zone)
+            .with_query_type(RecordType::NS)
+            .with_section(MockResponseSection::Additional),
```

Comment:
These builder calls look unusual here, should they be on the A record below? Same with the next NS record.